scrapy/extensions/feedexport.py

```diff
@@ -93,12 +93,29 @@ def store(self, file):
 
 class S3FeedStorage(BlockingFeedStorage):
 
-    def __init__(self, uri):
-        from scrapy.conf import settings
+    def __init__(self, uri, access_key=None, secret_key=None):
+        # BEGIN Backwards compatibility for initialising without keys (and
+        # without using from_crawler)
+        no_defaults = access_key is None and secret_key is None
+        if no_defaults:
+            from scrapy.conf import settings
+            if 'AWS_ACCESS_KEY_ID' in settings or 'AWS_SECRET_ACCESS_KEY' in settings:
+                import warnings
+                from scrapy.exceptions import ScrapyDeprecationWarning
+                warnings.warn(
+                    "Initialising `scrapy.extensions.feedexport.S3FeedStorage` "
+                    "without AWS keys is deprecated. Please supply credentials or "
+                    "use the `from_crawler()` constructor.",
+                    category=ScrapyDeprecationWarning,
+                    stacklevel=2
+                )
+                access_key = settings['AWS_ACCESS_KEY_ID']
```

Comment:
Just checking: given that the operator here is an `or`, a `KeyError` might raise if one of the keys is missing, I guess that's expected? User code will break with a clear message I think.

Reply:
Settings object returns None if a key is missing (see https://github.com/scrapy/scrapy/blob/258121db044c212493560dc5529b4355b148dae8/scrapy/settings/__init__.py#L93)

Reply:
In that case, I _think_ this might still break in the `boto` side if one of `access_key`/`secret_key` is not set either in the settings or in `uri`, the error message there will probably be clear enough.